### PR TITLE
Progressive Beam now are saved locally as Beam = 2

### PR DESCRIPTION
**Pull request by UltiNaruto**

### CODE REVIEW

File: Container.py [hunk 1]
```diff
@@ -167,35 +182,35 @@ def add(output_register: GeneralRegister, input_register1: GeneralRegister, inpu
     if not progressive_beams:
         return []
     cstate_manager_global = GAMES[game_version]["cstate_manager_global"]
-    charge_beam_offset = 0x7F
```
Comment (hesto2) [inline Container.py:170]: Is this the same address but essentially reading the byte rather than the bit?

File: Container.py [hunk 2]
```diff
@@ -167,35 +182,35 @@ def add(output_register: GeneralRegister, input_register1: GeneralRegister, inpu
...
         return []
     cstate_manager_global = GAMES[game_version]["cstate_manager_global"]
-    charge_beam_offset = 0x7F
+    charge_beam_offset = 0x7C
     instructions: List = [
         # Step 0: Get the player state address
         *_load_player_state_to_r6(game_version),
 
         # Step 1: Get the current beam from the player state
-        lbz(r5, 0xB, r6),  # Load the current beam value
```
Comment (hesto2) [inline Container.py:176]: Same here, I assume?

File: Container.py [hunk 3]
```diff
@@ -167,35 +182,35 @@ def add(output_register: GeneralRegister, input_register1: GeneralRegister, inpu
...
+        lwz(r5, 0x8, r6),          # Load the current beam value
 
-        # Step 2: Read the value at the progressive beam address
-        lis(r7, GAMES[game_version]["PROGRESSIVE_BEAM_ADDRESS"] >> 16),  # Load upper 16 bits of progressive beam address
-        ori(r7, r7, GAMES[game_version]["PROGRESSIVE_BEAM_ADDRESS"] & 0xFFFF),  # Load lower 16 bits of progressive beam address
-        add(r7, r7, r5),  # Add 0 to the address (no offset
-        lbz(r8, 0, r7),  # Load the value at the progressive beam address
+        # Step 2: Read the value at the beam address in CPlayerState::PowerUps[41]
```
Comment (hesto2) [inline Container.py:193]: Can you give just a general description of what's going on here? I see the comments so I have a vague idea of the 'what', but I'm not completely following the 'why' and I definitely want to make sure I understand this well enough to maintain it solo if necessary before merging it in 😂

File: MetroidPrimeInterface.py [hunk 4]
```diff
@@ -186,8 +186,19 @@ def __init__(self, logger) -> None:
     def give_item_to_player(self, item_id: int, new_amount: int, new_capacity: int, ignore_capacity: bool = False):
         """Gives the player an item with the specified amount and capacity"""
         if item_id in custom_charge_id_to_beam:
-            charge_beam = custom_charge_id_to_beam[item_id]
-            self.set_progressive_beam_charge_state(charge_beam, True)
```
Comment (hesto2) [inline MetroidPrimeInterface.py:190]: Could you encapsulate this all inside the  existing function instead? It seems like there could probably be a way to reduce a lot of this down into a similarly sized function

Comment (hesto2) [reply MetroidPrimeInterface.py:190]: I think we'd also want to reimplement the `get_progressive_beam_charge_state` function and remove the `"PROGRESSIVE_BEAM_ADDRESS"`

File: Container.py (hunk 1 above)
Comment (UltiNaruto) [reply Container.py:170]: you mean reading as int rather than byte?

File: Container.py (hunk 3 above)
Comment (UltiNaruto) [reply Container.py:193]: Reading at CPlayerState::PowerUps[beam].x4_capacity.

Where 0 is no beam
Where 1 is beam
Where 2 is charge beam

File: MetroidPrimeInterface.py (hunk 4 above)
Comment (UltiNaruto) [reply MetroidPrimeInterface.py:190]: I will edit the PR with the changes

File: Container.py (hunk 1 above)
Comment (UltiNaruto) [reply Container.py:170]: and yes it is the same thing we're reading from

File: MetroidPrimeInterface.py [hunk 5]
```diff
@@ -387,27 +380,44 @@ def _save_message_to_memory(self, message: str):
...
-        elif charge_beam == SuitUpgrade.Plasma_Charge_Beam:
-            offset = 3
-        return offset
-
-    def set_progressive_beam_charge_state(self, beam: SuitUpgrade, state: bool):
-        offset = self.__get_progressive_beam_charge_offset(beam)
-        state = b"\x01" if state else b"\x00"
-        self.dolphin_client.write_address(GAMES[self.current_game]["PROGRESSIVE_BEAM_ADDRESS"] + offset, state)
-
-    def get_progressive_beam_charge_state(self, beam: SuitUpgrade) -> bool:
```
Comment (hesto2) [inline MetroidPrimeInterface.py:407]: Nice 👍

File: MetroidPrimeInterface.py [hunk 6]
```diff
@@ -218,23 +212,22 @@ def set_cosmetic_suit_by_id(self, item_id: int):
     def check_for_new_locations(self):
         pass
 
-    def get_item(self, item_id: int) -> InventoryItemData:
```
Comment (hesto2) [inline MetroidPrimeInterface.py:221]: I think this overload is actually still referenced elsewhere in the client. You'll want to run it locally (or build the apworld locally using `bash/build.build.sh` and install it) to verify the client is able to give/receive the progressive items correctly.
When I ran the client with the game I was getting a bunch of errors about some missing attribtutes and functions

Comment (UltiNaruto) [reply MetroidPrimeInterface.py:221]: do you have the lines where the errors are?

Comment (UltiNaruto) [reply MetroidPrimeInterface.py:221]: oh I forgot to handle the None type

File: ClientReceiveItems.py [hunk 7]
```diff
@@ -81,7 +81,7 @@ async def handle_disable_gravity_suit(ctx: 'MetroidPrimeContext', current_items:
 
 async def handle_receive_missiles(ctx: 'MetroidPrimeContext', current_items: dict[str, InventoryItemData]):
     # Slot data is required for missiles + Power Bombs so we can check if launcher / main pb are enabled
-    if ctx.slot_data is not None:
```
Comment (hesto2) [inline ClientReceiveItems.py:84]: Can we separate these optimizations into a separate PR? I want to be able to test them separately

Comment (hesto2) [reply ClientReceiveItems.py:84]: Ah I thought they were more expansive, I can test these. If you could use the SuitUpgrade enum value instead of strings for these that'd be 🔥

Comment (hesto2) [reply ClientReceiveItems.py:84]: Ah disregard that as well, I can just add that post merge. I saw that these were already not using the enum
